### DISCUSSION TITLE
"window is not defined" in useEffect

### DISCUSSION THREAD

**Discussion by alekseykurylev** [vercel/next.js#67597] [Help] [answered]
### Summary

I am using a package that uses the `window` object. Initialization of this package is performed in the client component inside the `useEffect`. But I still get the error `window is not defined`. Adding an additional check `if (typeof window !== 'undefined')` does not fix the error.

```tsx
'use client'

import {useEffect, useState} from "react";
import { CryptoPro, CertListItemInterface } from 'ruscryptojs';

export default function Home() {
  const [certs, setCerts] = useState<CertListItemInterface[]>([]);

  useEffect(() => {
    const cryptopro = new CryptoPro();
    const listCertificates = async () => {
      await cryptopro.init();
      const list = await cryptopro.listCertificates();
      setCerts(list)
    };
    listCertificates().catch((error) => {
      console.log(error);
    });
  }, []);

  return <div>{certs.map((item) => <div key={item.id}>{item.name}</div>)}</div>
}
```

<img width="697" alt="image" src="https://github.com/vercel/next.js/assets/8526127/094a5e3c-0e91-45cf-a32f-677946b946d4">

https://github.com/Aleksandr-ru/RusCryptoJS/blob/master/src/CryptoPro/index.js

next 14.2.3
react 18.3.1

**Comment by BorhanSaflo** [comment]
Move the component into its own file and import it [dynamically](https://nextjs.org/docs/app/building-your-application/optimizing/lazy-loading#nextdynamic) with server-side rendering (SSR) disabled. This will ensure the component is not pre-rendered.

Example:

**component.tsx**
```ts
"use client";

import { useEffect, useState } from "react";
import { CryptoPro, CertListItemInterface } from 'ruscryptojs';

export default function Component() {
  const [certs, setCerts] = useState<CertListItemInterface[]>([]);

  useEffect(() => { 
    const cryptopro = new CryptoPro();
    const listCertificates = async () => {
      await cryptopro.init();
      const list = await cryptopro.listCertificates();
      setCerts(list);
    };
    listCertificates().catch((error) => {
      console.error(error);
    });
  }, []);

  return (
    <div>
      {certs.map((item) => (
        <div key={item.id}>{item.name}</div>
      ))}
    </div>
  );
}
```

**page.tsx**
```ts
import dynamic from "next/dynamic";

const Component = dynamic(() => import("./component"), { ssr: false });

export default function Home() {
  return (
    <Component />
  );
}
```

**Reply by alekseykurylev** [reply]
'next/dynamic' was used, the error remains.

**Reply by BorhanSaflo** [reply, marked as answer]
I've set up a [CodeSandbox](https://codesandbox.io/p/devbox/funny-surf-2662yv?file=%2Fapp%2Fcomponent.tsx%3A20%2C8) to test the example above. It ran without any issues.
